flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/ComponentMetricGroup.java

```diff
@@ -57,11 +57,12 @@ public ComponentMetricGroup(MetricRegistry registry, String[] scope, P parent) {
 		if (variables == null) { // avoid synchronization for common case
 			synchronized (this) {
 				if (variables == null) {
-					variables = new HashMap<>();
-					putVariables(variables);
+					Map<String, String> tmpVariables = new HashMap<>();
+					putVariables(tmpVariables);
 					if (parent != null) { // not true for Job-/TaskManagerMetricGroup
-						variables.putAll(parent.getAllVariables());
+						tmpVariables.putAll(parent.getAllVariables());
 					}
+					variables = tmpVariables;
```

Comment:
Do not need any test that verify this change ?

Reply:
We _could_ write a test, but it would be heavily tied to the implementation and is thus a bit questionable.

Reply:
This looks potentially still broken to me because it does double-checked locking, which is a defective anti-pattern.

Reply:
To be clear, this is not exactly about the fixed lines, but the line that is commented with "optimization for common case" is the remaining problem.

Reply:
EDIT removed my previous comment about double-checked locking because I saw that the variable is `volatile`